src/families/tron/react.js

```diff
@@ -68,3 +88,166 @@ export const formatVotes = (
+
+  const SR = useMemo(
+    () =>
+      superRepresentatives.map((sr, rank) => ({
+        sr,
+        name: sr.name,
+        address: sr.address,
+        rank: rank + 1,
+        isSR: rank <= SR_THRESHOLD
+      })),
+    [superRepresentatives]
+  );
+
+  const sortedVotes = useMemo(
+    () =>
+      SR.filter(({ address }) => initialVotes.includes(address)).concat(
+        SR.filter(({ address }) => !initialVotes.includes(address))
+      ),
+    [SR, initialVotes]
+  );
+
+  const sr = useMemo(
+    () => (search ? SR.filter(searchFilter(search)) : sortedVotes),
+    [search, SR, sortedVotes]
+  );
+
+  return sr;
+}
+
+/** Tron vote flow state reducer */
+function votesReducer(state: State, { type, address, value }: Action) {
+  switch (type) {
+    case "updateVote": {
+      const voteCount = value
+        ? parseInt(Number(value.replace(/[^0-9]/g, "")))
+        : 0;
+      const currentVotes = Object.values({
+        ...state.votes,
+        [address]: voteCount
+      }).filter(Boolean);
+
+      const votes = {
+        ...state.votes,
+        [address]:
+          voteCount <= 0 || currentVotes.length > SR_MAX_VOTES ? 0 : voteCount
+      };
+
+      const votesUsed = Object.values(votes).reduce(
+        (sum, count) => sum + Number(count),
+        0
+      );
+
+      return {
+        ...state,
+        votes,
+        votesUsed,
+        votesSelected: Object.values(votes).filter(Boolean).length,
+        max: Math.max(0, state.votesAvailable - votesUsed)
+      };
+    }
+    case "resetVotes": {
+      const { initialVotes, votesAvailable } = state;
+      const votesUsed = Object.values(initialVotes).reduce(
+        (sum, voteCount) => sum + Number(voteCount),
+        0
+      );
+      return {
+        ...state,
+        votes: initialVotes,
+        votesUsed,
+        votesSelected: Object.keys(initialVotes).length,
+        max: Math.max(0, votesAvailable - votesUsed)
+      };
+    }
+    case "clearVotes": {
+      return {
+        ...state,
+        votes: {},
+        votesUsed: 0,
+        votesSelected: 0,
+        max: state.votesAvailable
+      };
+    }
+    default:
+      return state;
+  }
+}
+
+/** Vote flow init state */
+function initState(initialVotes, tronResources): State {
+  const votes = initialVotes.reduce(
+    (sum, { voteCount, address }) => ({ ...sum, [address]: voteCount }),
+    {}
+  );
+  const votesAvailable = tronResources ? tronResources.tronPower : 0;
+  const votesUsed = Object.values(votes).reduce(
+    (sum, voteCount) => sum + Number(voteCount),
+    0
+  );
+
+  return {
+    votes,
+    votesAvailable,
+    votesUsed,
+    votesSelected: initialVotes.length,
+    max: Math.max(0, votesAvailable - votesUsed),
+    initialVotes: votes
+  };
+}
+
+/** Hook to retrieve and update voting flow state */
+export function useVotesReducer(
+  votes: Vote[],
+  onChangeVotes: (votes: Vote[]) => void,
```

Comment:
having both an external state and an internal state of "votes" feels like it's an antipattern; there is something to improve here. there should be no "onChangeVotes" I feel. `votes` param should be called `initialVotes` if it's truely what's happening here.

Reply:
initially i wanted to debounce the update of the transaction votes and have an easier to use data structure for this but updating the transaction after each update doesn't seem to make the whole thing lag anymore :man_shrugging:

Reply:
we should `useDebounce` then. because in your code, it's never calling onChangeVotes

Reply:
the useEffect with a debounce does that, it's weird but it works i'll try with the fixed use debouncedCallback on the lld side without the debouncing here :+1:

Reply:
ok i understand the code, but it's not optimal and it's very dangerous: if you have a lot of re-render happening, the debouncing will never be called, it's only called when there is no re-render but you don't know that